File: controller/src/main/java/io/github/oxi1224/logging/Logger.java
package io.github.oxi1224.logging;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class Logger {
  public static final String ANSI_RESET = "\033[0m";
  public static final String ANSI_RED = "\033[38;5;160m";
  public static final String ANSI_GREEN = "\033[38;5;157m";
  public static final String ANSI_YELLOW = "\033[38;5;220m";
  public static final String ANSI_GRAY = "\033[38;5;251m";
  public static final String ANSI_WHITE = "\033[37m";

  private static String logFilePath = "";
  private static DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd | HH:mm:ss");
  private String filePath;
  private Class<?> clazz;

  private Logger(Class<?> clazz, String filePath) {
    this.clazz = clazz ;
    this.filePath= filePath;
  }

  public void setLogFilePath(String logFilePath) {
    Logger.logFilePath = logFilePath;
  }

  public static Logger getLogger(Class<?> clazz) {
    return getLogger(clazz, "log.txt");
  }

  public static Logger getLogger(Class<?> clazz, String fileName) {
    String filePath = logFilePath + fileName;
    File logFile = new File(filePath);
    if (!logFile.exists()) {
      try {
        logFile.createNewFile();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
    if (!logFile.canWrite()) {
      throw new IllegalArgumentException(String.format("File %s cannot be written to", filePath));
    }
    return new Logger(clazz, filePath);
  }

  public void info(String text) {
    write("[INFO] " + text, ANSI_WHITE);
  }

  public void infof(String text, Object... fargs) {
    info(String.format(text, fargs)); 
  }

  public void debug(String text) {
    write("[DEBUG] " + text, ANSI_GRAY);
  }

  public void debugf(String text, Object... fargs) {
    debug(String.format(text, fargs)); 
  }

  public void warn(String text) {
    write("[WARN] " + text, ANSI_YELLOW);
  }

  public void warnf(String text, Object... fargs) {
    warn(String.format(text, fargs));
  }

  public void error(String text) {
    write("[ERROR] " + text, ANSI_RED);
  }

  public synchronized void error(String text, Throwable throwable) {
    StringBuilder sb = new StringBuilder();
    if (!text.isBlank()) sb.append(text + " ");
    sb.append(throwable.getMessage());
    sb.append(" [");
    sb.append(clazz.getPackageName() + ".");
    sb.append(clazz.getName() + ".");
    sb.append(throwable.getStackTrace()[0].getMethodName());
    sb.append("]: ");
    sb.append(throwable.getClass().getPackageName() + ".");
    sb.append(throwable.getClass().getName());
    sb.append("\n\t");
    StringWriter sw = new StringWriter();
    PrintWriter pw = new PrintWriter(sw);
    throwable.printStackTrace(pw);
    sb.append(sw.toString());
    sb.delete(sb.length() - 2, sb.length()); // Remove the last newline

    error(sb.toString());
  }

  public void error(Throwable throwable) {
    error("", throwable);
  }

  public void errorf(String text, Object... fargs) {
    error(String.format(text, fargs)); 
  }

  public void errorf(String text, Throwable throwable, Object... fargs) {
    error(String.format(text, fargs), throwable);
  }

  private synchronized void write(String text, String ansiColor) {
    try {
      FileWriter w = new FileWriter(filePath, true);
      String dateStr = "[" + LocalDateTime.now().format(dateFormatter) + "]  ";
      w.write(dateStr + text + "\n");
      System.out.println(ANSI_GREEN + dateStr + ANSI_RESET + ansiColor + text + ANSI_RESET);
      w.close();
    } catch (IOException e) {
      e.printStackTrace();  
    }
  }
}

File: client/src/logger.cpp
#include <ctime>
#include <fstream>
#include <iostream>
#include "logger.hpp"

Logger::Logger(std::string logID) {
  m_logID = logID;
  if (log_to_file == true) m_file = std::ofstream(m_logFileName, std::fstream::app);
}

Logger::Logger(std::string logID, std::string filename) {
  m_logID = logID;
  m_logFileName = filename;
  if (log_to_file == true) m_file = std::ofstream(m_logFileName, std::fstream::app);
}

Logger::~Logger() {
  if (m_file.is_open()) m_file.close();
}

std::string Logger::getTimeStamp() {
	std::time_t rawtime;
  time(&rawtime);
	struct tm timeinfo;
  localtime_s(&timeinfo, &rawtime);
	std::string strHr = std::to_string(timeinfo.tm_hour);
	std::string strMin = std::to_string(timeinfo.tm_min);
	std::string strSec = std::to_string(timeinfo.tm_sec);

	if (strHr.length() != 2) {
		strHr = "0" + strHr;
	}
	if (strMin.length() != 2) {
		strMin = "0" + strMin;
	}
	if (strSec.length() != 2) {
		strSec = "0" + strSec;
	}

	std::string timestamp = "[";
	timestamp += strHr;
	timestamp += ":";
	timestamp += strMin;
	timestamp += ":";
	timestamp += strSec;
	timestamp += "] ";
  return timestamp;
}

void Logger::write(const std::string& txt, const std::string& ansi_color, const std::string& log_type) {
  std::string log_info = "[" + m_logID + "/" + log_type + "] ";
  std::cout << ANSI_GREEN << getTimeStamp() << ANSI_RESET << ansi_color << log_info << txt << ANSI_RESET << std::endl;
  if (log_to_file && !m_file.is_open()) {
    m_file = std::ofstream(m_logFileName, std::fstream::app);
  }
  if (log_to_file) m_file << getTimeStamp() << log_info << txt << std::endl;
}

void Logger::log(const std::string& str) {
  write(str, ANSI_WHITE, "INFO");
}

void Logger::warn(const std::string& str) {
  write(str, ANSI_YELLOW, "WARN");
}

void Logger::error(const std::string& str) {
  write(str, ANSI_RED, "ERROR");
}

void Logger::debug(const std::string& str) {
  write(str, ANSI_GRAY, "DEBUG");
}

